xline/src/server/lease_server.rs

```diff
@@ -133,13 +145,28 @@ where
     }
 
     /// Handle keep alive at leader
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select!
     async fn leader_keep_alive(
         &self,
         mut request_stream: tonic::Streaming<LeaseKeepAliveRequest>,
+        mut shutdown_listener: shutdown::Listener,
     ) -> Pin<Box<dyn Stream<Item = Result<LeaseKeepAliveResponse, tonic::Status>> + Send>> {
         let lease_storage = Arc::clone(&self.lease_storage);
         let stream = try_stream! {
-            while let Some(keep_alive_req) = request_stream.message().await? {
+           loop {
+                let keep_alive_req: LeaseKeepAliveRequest = tokio::select! {
+                    _ = shutdown_listener.wait_self_shutdown() => {
+                        debug!("Lease keep alive shutdown");
+                        break;
+                    }
+                    res = request_stream.message() => {
+                        if let Ok(Some(keep_alive_req)) = res {
+                            keep_alive_req
+                        } else {
+                            break;
+                        }
+                    }
+                };
```

Comment:
Could the future be safely canceled without the shutdown listener?

Reply:
Here is a special case, the server must actively disconnect from the client when it shutdown, otherwise `tonic::Server `will not stop.